### ISSUE TITLE
worker may become a persistent key in redis

### ISSUE THREAD

**Issue by valentinogeron** [rq/rq#1276]
The worker is being represented as redis key (`rq:worker:<worker_name>`).
This key is a volatile key - on creation and on heartbeat, expire command is being used.
However, there are some cases (e.g. `set_state` function) where this key is modified but there is no check for the key existence.
It may result in a persistence key. For example:
``` 
127.0.0.1:6379> HGETALL rq:worker:my_worker:1
1) "state"
2) "idle"
127.0.0.1:6379> TTL rq:worker:my_worker:1
(integer) -1
````
Now, when we have a persistent key, if the worker will exit unexpectedly, without calling to `register_death` function, a new worker with the same name will not be able to start up, because a key which represents this worker already exists and it does not contain `death` field.

**Comment by selwin** [comment]
I made some changes in this PR so that worker keys are expired less aggressively. I hope this helps.

Do you mind taking the latest master branch for a spin to see if it solves your problem?

I looked around the source code for places where key expiration could potentially be a problem and found that we could perhaps add a heartbeat in `monitor_work_horse()` after the work ends. But before that, let's see if this change helps so we don't need to add an extra Redis call for each job cycle.

**Comment by valentinogeron** [comment]
I have looked at your fix, and I think that this approach will cover the bug rather than solving it.

For `String` type in redis, we have an optional parameter for `SET` command which is `XX`. This means only set the key if it already exists. But, unfortunately, there is no equivalent for `Hash` type.

I think about 2 possible options:
1. using `LUA` script, in which we first check if the hash exists and only then we update its fields.
2. every time that we update the worker state in redis, we update all the relevant fields (using MULTI..EXEC)

In case that we do not care about the current data in the redis, IMHO, option 2 is more suitable.
@selwin - WDYT?

**Comment by selwin** [comment]
> I have looked at your fix, and I think that this approach will cover the bug rather than solving it.

The bug we're talking about here is the worker's Redis key expiring when the worker is still alive, right? So tuning the TTL to be less agressive is a direct way to address this. A nuclear option would be to make the worker keys permanent, but I don't think that's the approach we want to take.

> every time that we update the worker state in redis, we update all the relevant fields
I feel like doing this would be horribly inefficient and to borrow your term, this is just a workaround instead of fixing the actual bug (worker key expiring when worker is still alive) :).

**Comment by valentinogeron** [comment]
What about the first option (use LUA script) ?

**Comment by selwin** [comment]
Only if it's still a problem after the patch. I'd also want to know where in the worker lifecycle the key expires. Checking for key existence everytime there's an update is inelegant and would still result in an extra Redis operation, even if lua script is used.

**Comment by selwin** [comment]
I have released 1.4.3 on PyPI. I'm gonna close this issue for now. Please re-open if this is still an issue.

**Comment by guybe7** [comment]
hi @selwin

i'm also affected by this issue

the fix proposed (and implemented) for this issue (longer TTL) just makes the problem more unlikely to happen but doesn't eliminate it completely.
the root cause is that not all operations verify the key is still volatile. moreover even if you do first EXPIRE the key and then modify, there's no guarantee that by the time of the modification the key hasn't expired already (imagine really slow machine running the app while a fast machine running Redis). for example, in `heartbeat`:
```
        connection.expire(self.key, timeout)
        connection.hset(self.key, 'last_heartbeat', utcformat(utcnow()))
```
in extreme cases, the key could be already expired when you call HSET.

i'd like to suggest a slightly different approach: have an additional key per worker just for heartbeat purposes (e.g. `rq:worker:my_worker:1:heartbeat`). the only action on this key is SETEX which is done in `worker.heartbeat()`. the key holding the worker state is not volatile.
in `register_birth`, instead of checking the existence of the `death` field, just check the existence of the heartbeat key.

please let me know if you think this approach is solid.
